pkg/controller/plan/adapter/vsphere/builder.go

```diff
@@ -1034,32 +1045,213 @@ func (r *Builder) LunPersistentVolumeClaims(vmRef ref.Ref) (pvcs []core.Persiste
+					} else {
+						pvc.Annotations = annotations
+					}
+					pvc.Annotations[planbase.AnnDiskSource] = r.baseVolume(disk.File)
+					pvc.Annotations["copy-offload"] = r.baseVolume(disk.File)
+					pvcs = append(pvcs, &pvc)
+
+					vp := api.VSphereXcopyVolumePopulator{
+						ObjectMeta: metav1.ObjectMeta{
+							Name:      commonName,
+							Namespace: namespace,
+							Labels:    labels,
+						},
+						Spec: api.VSphereXcopyVolumePopulatorSpec{
+							VmdkPath:             disk.File,
+							TargetPVC:            commonName,
+							StorageVendorProduct: storageVendorProduct,
+							SecretRef:            secretName,
+						},
+					}
+
+					// Ensure a Secret combining Vsphere and Storage secrets
+					err = r.mergeSecrets(secretName, namespace, storageVendorSecretRef, r.Source.Provider.Namespace)
+					if err != nil {
+						return nil, fmt.Errorf("failed to merge secrets for popoulators", err)
+					}
+					// TODO should we handle if already exists due to re-entry? if the former
+					// reconcile was successful in creating the pvc but failed after that, e.g when
+					// creating the volumepopulator resouce failed
+					r.Log.Info("Creating pvc", "pvc", pvc)
+					err = r.Client.Create(context.TODO(), &pvc, &client.CreateOptions{})
+					if err != nil {
+						// ignore if already exists?
+						return nil, err
+					}
+					r.Log.Info("Creating volumepopulator", "volumepopulator", vp)
+					err = r.Client.Create(context.TODO(), &vp, &client.CreateOptions{})
+					if err != nil {
+						return nil, err
+					}
+				}
+			}
+		}
+	}
+
+	return pvcs, nil
+
 }
 
 func (r *Builder) PrePopulateActions(c planbase.Client, vmRef ref.Ref) (ready bool, err error) {
 	err = planbase.VolumePopulatorNotSupportedError
 	return
 }
 
-func (r *Builder) PopulatorTransferredBytes(persistentVolumeClaim *core.PersistentVolumeClaim) (transferredBytes int64, err error) {
-	err = planbase.VolumePopulatorNotSupportedError
+func (r *Builder) PopulatorTransferredBytes(pvc *core.PersistentVolumeClaim) (transferredBytes int64, err error) {
+	vmdkKey := pvc.Labels["vmdkKey"]
+	populatorCr, err := r.getVolumePopulator(vmdkKey)
+	if err != nil {
+		return
+	}
+
+	progressPercentage, err := strconv.ParseInt(populatorCr.Status.Progress, 10, 64)
+	if err != nil {
+		r.Log.Error(err, "Couldn't parse the progress percentage.", "pvcName", pvc.Name, "progressPercentage", progressPercentage)
+		transferredBytes = 0
+		err = nil
+		return
+	}
+
+	pvcSize := pvc.Spec.Resources.Requests["storage"]
+	transferredBytes = (progressPercentage * pvcSize.Value()) / 100
+
 	return
 }
 
+func (r *Builder) getVolumePopulator(vmdkKey string) (api.VSphereXcopyVolumePopulator, error) {
+	list := api.VSphereXcopyVolumePopulatorList{}
+	err := r.Destination.Client.List(context.TODO(), &list, &client.ListOptions{
+		Namespace: r.Plan.Spec.TargetNamespace,
+		LabelSelector: labels.SelectorFromSet(map[string]string{
+			"migration": string(r.Migration.UID),
+			"vmdkKey":   vmdkKey,
```

Comment:
The vmdkKey by itself is not unique enough, you need to also specify the VM as you can have multiple disks in the same plan with the same disk key.

Reply:
ack